File: 2_SoSe_22/Systemprogrammierung/src/aufgabe06/bin/notenspiegel.cpp
/*
 * notenspiegel.cpp
 *
 * Liest die Namen von Faechern mit den zugehoerigen Noten in eine
 * verkettete Liste ein und gibt dann einen Notenspiegel aus.
 *
 * Autor: H.Drachenfels
 * Version: 19.12.2019
 */
#include "../lib/benotung.h"
#include "../lib/fachnote.h"
#include "../lib/fachnoten_liste.h"
#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <cstdio>
#include <string>
#include <numeric>
#include <algorithm>
#include <forward_list>

namespace
{
std::ostream& operator<<(std::ostream& os, const benotung& n);
void delete_fachnote(fachnote *);
}

int read_grade(int& note);

int main()
{
    fachnoten_liste liste(delete_fachnote); // leere Liste
    //std::forward_list<std::shared_ptr<fachnote>> liste;
    //--------------------------------------------- Notenspiegel einlesen
    std::cerr << "Fach und Note zwischen "
              << benotung::beste.int_value()
              << " und "
              << benotung::schlechteste.int_value()
              << " eingeben (Ende mit Strg-D):\n";

    while (true)
    {
        try
        {
            // Fach und Note einlesen:
            std::string fach;
            if (!(std::cin >> fach))
            {
                break;
            }

            int note;
            int res = read_grade(note);
            if (res == -1)
            {
                std::cerr << "Fach " << fach << " ohne Note ignoriert!\n";
                break;
            }
            else if (res == 0)
            {
                std::string s;
                std::cin >> s;
                std::cerr << "Eingabefehler: " << s << '\n';
                continue;
            }

            int note2;
            if (read_grade(note2) != 1)
            {
                // neue Fachnote in Notenliste eintragen:
                auto* f = new fachnote{fach, benotung{note}};
                liste.insert(f);
            }
            else
            {
                // neue Fachnote in Notenliste eintragen:
                auto* f = new fachnote{fach, benotung{note, note2}};
                liste.insert(f);
            }
        }
        catch (std::invalid_argument& x)
        {
            std::cerr << "Eingabefehler: " << x.what() << '\n';
            continue;
        }
    }

    //--------------------------------------------- Notenspiegel ausgeben
    std::size_t max_laenge = std::accumulate(liste.begin(), liste.end(), 0, [](size_t a, fachnote* b)
    {
        if (a < b->fach.length())
        {
            return b->fach.length();
        }
        return a;
    });

    std::cout << "\nNOTENSPIEGEL\n";
    for (const auto& f : liste)
    {
        std::cout << std::setw(static_cast<int>(max_laenge)) << std::left << f->fach << '\t'
                  << f->note << '\t';
        if (f->note == benotung::beste)
        {
            std::cout << "mit Bestnote bestanden\n";
        }
        else if (f->note.ist_bestanden())
        {
            std::cout << "bestanden\n";
        }
        else
        {
            std::cout << "nicht bestanden\n";
        }
    }
}

int read_grade(int& note)
{
    if (!(std::cin >> note))
    {
        if (std::cin.eof())
        {
            return -1;
        }
        std::cin.clear();
        return 0;
    }
    return 1;
}

namespace
{
std::ostream& operator<<(std::ostream& os, const benotung& n)
{
    int note = n.int_value();
    os << note / 10 << ',' << note % 10;
    return os;
}

void delete_fachnote(fachnote *f)
{
    delete f;
}
}

File: 1_WiSe_21_22/Programmiertechnik_1/prog1-uebungen/aufgabe5/Notenspiegel.java
// Notenspiegel.java
package aufgabe5;
import java.util.NoSuchElementException;
import java.util.Scanner;

/**
 * Notenspiegel liest die Namen von F&auml;chern mit den zugeh&ouml;rigen Noten
 * in eine verkettete Liste ein und gibt dann einen Notenspiegel aus.
 * @author Christopher Mogler
 * @version 17.12.2021
 */
public final class Notenspiegel {
    private Notenspiegel() { }

    private static final Scanner SCANNER = new Scanner(System.in);

    /**
     * Prüft ob String ein Integer ist.
     * @param s zu prüfender String
     * @return true == String ist ein Integer
     */
    public static boolean isInteger(String s) {
        return isInteger(s, 10);
    }

    /**
     * Prüft ob String ein Integer ist.
     * @param s zu prüfender String
     * @param radix the radix.
     * @return true == String ist ein Integer
     */
    public static boolean isInteger(String s, int radix) {
        if (s.isEmpty()) {
            return false;
        }
        for (int i = 0; i < s.length(); i++) {
            if (i == 0 && s.charAt(i) == '-') {
                if (s.length() == 1) {
                    return false;
                } else {
                    continue;
                }
            }
            if (Character.digit(s.charAt(i), radix) < 0) {
                return false;
            }
        }
        return true;
    }

    /**
     * main ist der Startpunkt des Programms.
     * @param args wird nicht verwendet.
     */
    public static void main(String[] args) {
        final GradeList gradeList = new GradeList(); // leere Liste

        //--------------------------------------------- Notenspiegel einlesen
        System.err.printf(
            "Faecher mit Noten zwischen %d und %d eingeben "
            + "(Ende mit Strg-D):%n",
            Grade.BEST.intValue(), Grade.WORST.intValue());

        int maxLenSubject = 0;

        while (SCANNER.hasNextLine()) {
            String input = SCANNER.nextLine();
            String subject = null;
            Grade grade = null;

            try {
                //------------------------------------ Fach und Note einlesen

                /* (1) erst das Fach mit next() einlesen,
                             dann mit hasNextInt() pruefen,
                             ob die Note als ganze Zahl vorliegt,
                             und die Note je nachdem mit nextInt()
                             oder next() einlesen und schliesslich
                             in ein Wertobjekt verpacken */

                String[] ins = input.split(" ");
                subject = ins[0];

                if (maxLenSubject < subject.length()) {
                    maxLenSubject = subject.length();
                }

                for (int i = 1; i < ins.length && i < 3; ++i) {
                    String gr = ins[i];
                    Grade g = null;
                    if (isInteger(gr)) {
                        g = Grade.valueOf(Integer.parseInt(gr), subject);
                    } else {
                        g = Grade.valueOf(gr, subject);
                    }
                    if (grade == null) {
                        grade = g;
                    } else {
                        grade = Grade.valueOf(grade, g);
                    }
                }

                //--------------------- neue Fachnote in Notenliste eintragen

                /* (2) ein neues Fachnote-Objekt erzeugen
                             und am Listenanfang einfuegen */
                gradeList.insert(grade);

            } catch (IllegalArgumentException x) {
                System.err.printf("Eingabefehler: %s%n", x.getMessage());
                continue;
            } catch (NoSuchElementException x) {
                System.err.println("Fach ohne Note ignoriert!");
                break;
            }
        }

        //--------------------------------------------- Notenspiegel ausgeben

        /* (3) tabellarischen Notenspiegel
                     mit der Ueberschrift NOTENSPIEGEL ausgeben */

        System.out.println("\nNOTENSPIEGEL");

        for (Grade grade : gradeList) {
            System.out.println(grade.debug(maxLenSubject));
        }

    } // main
}

